runtime/src/drivers.rs

```diff
@@ -416,4 +428,55 @@ impl Drivers {
         drivers.cert_chain = cert_chain;
         Ok(())
     }
+
+    pub fn pl_context_threshold_exceeded(
```

Comment:
First two bullets done. For the third bullet, discussed in chat but the problem with making them member functions is that we get borrow checker errors since we mutably borrow drivers in other places in the code and it would take a large refactor to fix this.